CIP-0031/README.md

```diff
@@ -0,0 +1,210 @@
+
+### Accessing the datums of reference inputs
+
+Currently this proposal has somewhat limited utility, because datums in outputs are just hashes, and the spending party is required to find and provide the preimage themselves.
+Therefore the CIP-32 proposal for inline datums is complementary, as it allow UTXOs to carry the data itself rather than a hash.
+
+In the mean time (or if CIP-32 is not implemented), we still want users to be _able_ to provide the datums corresponding to datum hashes in reference outputs, so that at least it is possible to reference the data, albeit clumsily.
+The easiest solution here is to reuse the existing mechanism for providing pre-images of datum hashes which appear in outputs.
+
+Providing datum hash pre-images remains optional, since there are reasons to use reference inputs even without looking at the datum, e.g. to look at the value in an output.
+
+### Accessing the value locked in outputs
+
+The motivation of this proposal mainly requires looking at the _datums_ of outputs.
+But reference inputs allow us to do more: they let us look at the _value_ locked in an output (i.e. how much Ada or other tokens it contains).
+
+This is actually a very important feature.
+Since anyone can lock an output with any address, addresses are not that useful for identifying _particular_ outputs on chain, and instead we usually rely on looking for particular tokens in the value locked by the output.
+Hence, if a script is interested in referring to the data attached to a _particular_ output, it will likely want to look at the value is locked in the output.
+
+For example, an oracle provider would need to distinguish the outputs that they create (with good data) from outputs created by adversaries (with bad data). 
+They can do this with a token, so long as scripts can then see the token!
+
+### Hydra
+
+We want reference inputs to be usable inside Hydra heads.
+This raises a worry: since Hydra processes independent parts of the transaction graph in parallel, what happens if it accepts one transaction that references an output, and one that spends the output, but then the first transaction gets put into a checkpoint before the second?
+
+Fortunately, Hydra already has to deal with double-spend conflicts of this kind (although in the naive protocol this results in a decommit).
+This proposal simply introduces a new kind of conflict, a reference-spend conflict.
+Reference-spend conflicts should be handled by the same mechanism that is used to handle double-spend conflicts.
+
+### Controlling referencing
+
+One thing that a user might want to do is to control who can reference an output.
+For example, an oracle provider might want to only allow a transaction to reference a particular output if the transaction also pays them some money.
+
+Reference inputs alone do _not_ provide any way to do this.
```

Comment:
It seems like having some sort of access control over the datum in a reference input would be a very common use case. It might be good to mention how to do this or at least the rationale for this decision since to the uneducated (me) it seems like a simple matter of checking the witness on the reference input.

Reply:
I'll incorporate some of the discussion from this PR into the text.